packages/shorebird_cli/lib/src/commands/publish_command.dart

```diff
@@ -39,55 +55,197 @@ class PublishCommand extends ShorebirdCommand with ShorebirdConfigMixin {
+    final versionString = '${version.major}.${version.minor}.${version.patch}';
+
+    late final List<App> apps;
+    final fetchAppsProgress = logger.progress('Fetching apps');
     try {
-      final pubspecYaml = getPubspecYaml()!;
-      final shorebirdYaml = getShorebirdYaml()!;
-      final codePushClient = buildCodePushClient(
-        apiKey: session.apiKey,
-        hostedUri: hostedUri,
+      apps = (await codePushClient.getApps())
+          .map((a) => App(id: a.appId, displayName: a.displayName))
+          .toList();
+      fetchAppsProgress.complete();
+    } catch (error) {
+      fetchAppsProgress.fail('$error');
+      return ExitCode.software.code;
+    }
+
+    var app = apps.firstWhereOrNull((a) => a.id == shorebirdYaml.appId);
+    if (app == null) {
+      logger.info(
+        lightCyan.wrap("\nIt looks like this is a new app. Let's get started!"),
       );
-      logger.detail(
-        '''Deploying ${artifact.path} to ${shorebirdYaml.appId} (${pubspecYaml.version})''',
+      try {
+        app = await createApp();
+        addShorebirdYamlToProject(app.id);
+        addShorebirdYamlToPubspecAssets();
+        logger.info('''
+
+${lightGreen.wrap('🐦 Shorebird initialized successfully!')}
+
+✅ A shorebird app has been created.
+✅ A "shorebird.yaml" has been created.
+✅ The "pubspec.yaml" has been updated to include "shorebird.yaml" as an asset.''');
+      } catch (error) {
+        logger.err('$error');
+        return ExitCode.software.code;
+      }
+    }
+
+    logger.info(
+      '''
+
+${styleBold.wrap(lightGreen.wrap('🚀 Ready to publish a new patch!'))}
+
+📱 App: ${lightCyan.wrap(app.displayName)} ${lightCyan.wrap('(${app.id})')}
+📦 Release Version: ${lightCyan.wrap(versionString)}
+''',
+    );
+
+    final confirm = logger.confirm('Would you like to continue?');
+
+    if (!confirm) {
+      logger.info('Aborting.');
+      return ExitCode.success.code;
+    }
+
+    late final List<Release> releases;
+    final fetchReleasesProgress = logger.progress('Fetching releases');
+    try {
+      releases = await codePushClient.getReleases(
+        appId: app.id,
       );
-      final version = pubspecYaml.version!;
-      await codePushClient.createPatch(
+      fetchReleasesProgress.complete();
+    } catch (error) {
+      fetchReleasesProgress.fail('$error');
+      return ExitCode.software.code;
+    }
+
+    var release = releases.firstWhereOrNull(
+      (r) => r.version == versionString,
+    );
+
+    if (release == null) {
+      final createReleaseProgress = logger.progress('Creating release');
+      try {
+        release = await codePushClient.createRelease(
+          appId: app.id,
+          version: versionString,
+        );
+        createReleaseProgress.complete();
+      } catch (error) {
+        createReleaseProgress.fail('$error');
+        return ExitCode.software.code;
+      }
+    }
+
+    late final Patch patch;
+    final createPatchProgress = logger.progress('Creating patch');
```

Comment:
I've found most times that I have a late variable the block setting the late probably should be a function instead.

In this case, it seems like you want a function helper to deal with this "finally" behavior of managing the progress indicator?  e.g.

final Patch patch = logger.indeterminant("Creating patch", () { await codePushClient.createPatch(..));

Reply:
Yeah that could work I'll mess around with that API and potentially refactor in a subsequent PR 👍